src/game/server/gamemodes/instagib/tsmash/tsmash.cpp

```diff
@@ -0,0 +1,164 @@
+
+#include "tsmash.h"
+
+static int ColorToSixup(int Color6)
+{
+	return ColorHSLA(Color6).UnclampLighting(ColorHSLA::DARKEST_LGT).Pack(ColorHSLA::DARKEST_LGT7);
+}
+
+void CGameControllerTsmash::SetTeeColor(CPlayer *pPlayer)
+{
+	auto *pCharacter = pPlayer->GetCharacter();
+	if(!pCharacter)
+		return;
+	auto &LastHealth = m_aLastHealth[pPlayer->GetCid()];
+	if(LastHealth == pCharacter->Health())
+		return;
+	LastHealth = pCharacter->Health();
+	// Get color
+	ColorHSLA Color;
+	if(m_GameFlags & GAMEFLAG_TEAMS && pCharacter->Team() == TEAM_RED)
+	{
+		static constexpr ColorRGBA RED = ColorRGBA(1.0f, 0.5f, 0.5f);
+		Color = color_cast<ColorHSLA>(RED);
+		Color.h += (float)(10 - pCharacter->Health()) / 10.0f * 0.3f;
+	}
+	else if(m_GameFlags & GAMEFLAG_TEAMS && pCharacter->Team() == TEAM_BLUE)
+	{
+		static constexpr ColorRGBA BLUE = ColorRGBA(0.7f, 0.7f, 1.0f);
+		Color = color_cast<ColorHSLA>(BLUE);
+		Color.h -= (float)(10 - pCharacter->Health()) / 10.0f * 0.3f;
+	}
+	else
+	{
+		Color = ColorHSLA((float)(10 - pCharacter->Health()) / 10.0f * 0.75f + 0.25f, 1.0f, 0.5f);
+	}
+	// 0.6
+	auto Color6 = Color.Pack();
+	pPlayer->m_TeeInfos.m_UseCustomColor = 1;
+	pPlayer->m_TeeInfos.m_ColorBody = Color6;
+	pPlayer->m_TeeInfos.m_ColorFeet = Color6;
+	// 0.7
+	for(bool &UseCustomColor : pPlayer->m_TeeInfos.m_aUseCustomColors)
+		UseCustomColor = true;
+	const int Color7 = ColorToSixup(Color6);
+	for(int &PartColor : pPlayer->m_TeeInfos.m_aSkinPartColors)
+		PartColor = Color7;
+}
+
+CGameControllerTsmash::CGameControllerTsmash(class CGameContext *pGameServer, bool Teams) :
+	CGameControllerVanilla(pGameServer)
+{
+	m_pGameType = Teams ? "TTSmash2" : "TSmash2";
+	m_GameFlags = Teams ? GAMEFLAG_TEAMS : 0;
+
+	m_DefaultWeapon = WEAPON_HAMMER;
+
+	m_pStatsTable = Teams ? "ttsmash" : "tsmash";
+	m_pExtraColumns = nullptr;
+	m_pSqlStats->SetExtraColumns(m_pExtraColumns);
+	m_pSqlStats->CreateTable(m_pStatsTable);
+}
+
+CGameControllerTsmash::~CGameControllerTsmash() = default;
+
+int CGameControllerTsmash::OnCharacterDeath(class CCharacter *pVictim, class CPlayer *pKiller, int Weapon)
+{
+	if(!pKiller || pKiller->GetCid() == pVictim->GetPlayer()->GetCid())
+	{
+		const auto &LastToucher = pVictim->GetPlayer()->m_LastToucher;
+		if(LastToucher.has_value() && LastToucher->m_ClientId >= 0 && LastToucher->m_ClientId < (int)std::size(GameServer()->m_apPlayers) && GameServer()->m_apPlayers[LastToucher->m_ClientId])
+		{
+			OnKill(pVictim->GetPlayer(), GameServer()->m_apPlayers[LastToucher->m_ClientId], WEAPON_WORLD);
+			pVictim->Die(LastToucher->m_ClientId, WEAPON_WORLD);
+			return -1;
+		}
+	}
+	// PVP controller does not give score for "world" kills
+	return CGameControllerPvp::OnCharacterDeath(pVictim, pKiller,
+		Weapon == WEAPON_WORLD ? WEAPON_HAMMER : Weapon);
+}
+
+void CGameControllerTsmash::OnKill(class CPlayer *pVictim, class CPlayer *pKiller, int Weapon)
+{
+	// Taken from CGameControllerBaseFng::OnSpike
```

Comment:
If this is copy pasted code there should probably be one method both places can use. You can also create an issue for that does not have to happen in this pr but we should not forget it.